BASE/src/ServerTelemetryChannel/Implementation/TransmissionPolicy/AuthenticationTransmissionPolicy.cs

```diff
@@ -0,0 +1,124 @@
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void HandleTransmissionSentEvent(object sender, TransmissionProcessedEventArgs e)
+        {
+            if (this.Enabled)
+            {
+                HttpWebResponseWrapper httpWebResponseWrapper = e.Response;
+                if (httpWebResponseWrapper != null)
+                {
+                    if (httpWebResponseWrapper.StatusCode == ResponseStatusCodes.BadRequest)
+                    {
+                        // HTTP/1.1 400 Incorrect API was used - v2 API does not support authentication.
+                        // This indicates that the AI resource was configured for AAD, but SDK was not.
+                        // This is a configuration issue and is not recoverable. 
+                        this.ApplyHaltPolicy(e);
+                    }
+                    else if (httpWebResponseWrapper.StatusCode == ResponseStatusCodes.Unauthorized)
+                    {
+                        // HTTP/1.1 401 Unauthorized - please provide the valid authorization token
+                        // This indicates that the authorization token was either absent, invalid, or expired.
+                        // The root cause is not known and we should throttle retries.
+                        this.ApplyThrottlePolicy(e);
+                    }
+                    else if (httpWebResponseWrapper.StatusCode == ResponseStatusCodes.Forbidden)
+                    {
+                        // HTTP/1.1 403 Forbidden - provided credentials do not grant the access to ingest the telemetry into the component
+                        // This indicates the configured identity does not have permissions to publish to this resource.
+                        // This can be recovered if the user changes the AI Resource configuration.
+                        this.ApplyThrottlePolicy(e);
+                    }
+                }
+            }
+        }
+
+        private void ApplyHaltPolicy(TransmissionProcessedEventArgs e)
+        {
+            this.MaxSenderCapacity = 0;
+            this.MaxBufferCapacity = 0;
+            this.MaxStorageCapacity = null;
+
+            this.LogCapacityChanged();
+            this.Apply();
+
+            this.backoffLogicManager.ReportBackoffEnabled(e.Response.StatusCode);
+            this.Transmitter.Enqueue(e.Transmission);
+        }
+
+        private void ApplyThrottlePolicy(TransmissionProcessedEventArgs e)
+        {
+            this.MaxSenderCapacity = 0;
+            this.MaxBufferCapacity = null;
+            this.MaxStorageCapacity = null;
+
+            this.LogCapacityChanged();
+            this.Apply();
+
+            this.backoffLogicManager.ReportBackoffEnabled(e.Response.StatusCode);
+            this.Transmitter.Enqueue(e.Transmission);
+
+            // Ingestion service does not provide a retry value. We use our own here.
+            this.pauseTimer.Delay = TimeSpan.FromSeconds(30);
```

Comment:
i think we should do exponential backoffs, similar to how its used in other existing policies.

Reply:
Do we have a recommendation from Azure Identity on the retry timeframe?

Reply:
See: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/RetryOptions.cs

Exponential back-off. Maximum of 3 retries. Initial delay 0.8s; maximum delay 1 minute.

```csharp
        /// <summary>
        /// The maximum number of retry attempts before giving up.
        /// </summary>
        public int MaxRetries { get; set; } = 3;

        /// <summary>
        /// The delay between retry attempts for a fixed approach or the delay
        /// on which to base calculations for a backoff-based approach.
        /// </summary>
        public TimeSpan Delay { get; set; } = TimeSpan.FromSeconds(0.8);

        /// <summary>
        /// The maximum permissible delay between retry attempts.
        /// </summary>
        public TimeSpan MaxDelay { get; set; } = TimeSpan.FromMinutes(1);

        /// <summary>
        /// The approach to use for calculating retry delays.
        /// </summary>
        public RetryMode Mode { get; set; } = RetryMode.Exponential;

        /// <summary>
        /// The timeout applied to an individual network operations.
        /// </summary>
        public TimeSpan NetworkTimeout { get; set; } = TimeSpan.FromSeconds(100);
```

Also, the [following](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs#L207) is relevant (note the randomness applied). 

```csharp
    private TimeSpan CalculateExponentialDelay(int attempted)
    {
        return TimeSpan.FromMilliseconds(
            Math.Min(
                (1 << (attempted - 1)) * _random.Next((int)(_delay.TotalMilliseconds * 0.8), (int)(_delay.TotalMilliseconds * 1.2)),
                _maxDelay.TotalMilliseconds));
    }
```

Reply:
I think the `BackoffLogicManager` handles the exponential retries.
https://github.com/microsoft/ApplicationInsights-dotnet/blob/develop/BASE/src/ServerTelemetryChannel/Implementation/BackoffLogicManager.cs

Let me spend some time understanding how this works. 
In my local testing the retry is a fixed interval, so maybe I've configured something incorrectly.

Reply:
The `BackoffLogicManager` is not documented so I'm still trying to understand how this works.

It seems that the exponential backoff is triggered when it parses the retry value from the response header in `GetBackOffTimeInterval()`.
https://github.com/microsoft/ApplicationInsights-dotnet/blob/ce50b28c693fbb88dd749065039d9c1c57443ac9/BASE/src/ServerTelemetryChannel/Implementation/BackoffLogicManager.cs#L142-L175

THIS won't work for us because in the AAD scenarios, our Ingestion Service does not provide a retry header.

The only example I've found of this is in the `PartialSuccessTransmissionPolicy`:
https://github.com/microsoft/ApplicationInsights-dotnet/blob/ce50b28c693fbb88dd749065039d9c1c57443ac9/BASE/src/ServerTelemetryChannel/Implementation/TransmissionPolicy/PartialSuccessTransmissionPolicy.cs#L144-L155

I need more time to find a different way to use the existing code.

----

@cijothomas would it be possible to merge this as-is, without the exponential backoff? 
If so, this would unblock Beta3 and give me more time to work on a solution.

Reply:
You may be overthinking this. 

If it's just about AAD auth, really all you need to be concerned with is this:
If you get a response that indicates the token is invalid (401 Unauthorized), then you should go back to the TokenCredential and ask for a new token and _immediately_ retry. No need for any delay. If you get another 401, then fail the call (your credential is bad).

You can probably see why caching inside the TokenCredential is a hard to cope with. If you ask for a new token, but you get back the _exact_ same token (because it was cached and it isn't close enough to expiration), you're hosed. How do you deal with that? I think you just have to fail on the spot. 

For 403s (forbidden), the situation is subtle. That indicates that, while the token was good, you just don't have access (Metrics Publisher role is missing). It could be that you just added the role and it hasn't propagated to all regions yet. So, buffering on disk and retrying later might be the right thing -- but that's a long-cycle retry loop. Minutes, not seconds. It would be great if the SDK did that, but I wouldn't grumble if telemetry got dropped in this situation.

Reply:
Thanks, @pharring.
In this case, I'm advocating for avoiding the exponential backoff. And I'll change the retry to at least a minute.

I agree, I think data loss is going to be inevitable in most of these situations.